File: mundial/mundial.py
import sys
import copy
import itertools
import math as m

def check_perm(win, lose):
    for w, l in zip(win, lose):
        if not w.evale > l.efage:
            # print('Ivalid on ', w, l, w.evale, l.efage)
            return False
    return True


class Team:
    def __init__(self, text):
        self.name, agones, evale, efage = line.split(' ')
        self.agones = int(agones)
        self.efage = int(efage)
        self.evale = int(evale)

    __str__ = lambda self: self.name
    __repr__ = lambda self: self.name

class State:

    def __init__(self, w, l):
        self.winners = list(copy.deepcopy(w))
        self.losers = list(copy.deepcopy(l))

    def split_teams(self):
        new_winners, new_losers = [], []

        for w in self.winners:
            if w.agones == 1:
                new_losers.append(w)
            else:
                new_winners.append(w)
        return new_winners, new_losers

    def play_together(self):
        matches = []
        for i in range(len(self.winners)):
            w, l = self.winners[i], self.losers[i]
            w.agones -= 1
            match = (w, l, l.efage, l.evale)
            matches.append(match)
            w.evale -= l.efage
            w.efage -= l.evale
            self.winners[i] = w
        self.matches = matches



    def __repr__(self):
        return str(self.matches)

    def print_matches(self):
        for r in self.matches:
            w, l, x, y = r
            print('{}-{} {}-{}'.format(w,l,x,y))

    def is_final(self):
        if self.winners != []:
            return False
        else:
            x, y = self.losers
            if x.evale > x.efage:
                self.winners, self.losers = [x], [y]
            else:
                self.winners, self.losert = [y] , [x]
            self.play_together()
            # print(self.matches)
            assert(len(self.matches) == 1)
            return True


def solve(state):
    parent = {}
    stack = [state]
    parent[state] = None
    valid = False

    while stack != []:

        current = stack.pop()

        if current.is_final():
            # print('Final')
            # print(current)
            # print('true')
            valid = True
            return current, parent

        current.play_together()
        # print('Current is ', current.winners, current.losers)
        # print('Winners:')
        # for w in current.winners:
        #     print(w, w.evale, w.efage)
        # print('Losers')
        # for l in current.losers:
            # print(l, l.evale, l.efage)
        # print('Matches ', current.matches)

        new_winners, new_losers = current.split_teams()

        for winners_ in itertools.permutations(new_winners):
            if check_perm(winners_, new_losers):
                nxt = State(winners_, copy.copy(new_losers))
                stack.append(nxt)
                parent[nxt] = current

    return None, None






if __name__ == '__main__':
    with open(sys.argv[1]) as f:
        lines = f.read().splitlines()

    winners, losers = [], []

    # initial winners and losers
    for i, line in enumerate(lines):
        if i == 0:
            global N, K
            N = int(line)
            K = int(m.log2(N))
        else:
            t = Team(line)
            if t.agones == 1:
                losers.append(t)
            else:
                winners.append(t)

    # for w, l in zip(winners, losers):
    #     print(w, w.agones, w.efage, w.evale)
    #     print(l, l.agones, l.efage, l.evale)


    for winners_ in itertools.permutations(winners):
        if check_perm(winners_, losers):
            state = State(winners_, losers)
            final, parent = solve(state)
            if final and parent:
                current = final
                # print(final)
                # print('finished')
                # print('\nResults')
                while 42:
                    current.print_matches()
                    if parent[current] == None:
                        break
                    current = parent[current]
                exit()

File: mundial/mundial.pl
read_input(File, N, Teams) :-
    open(File, read, Stream),
    read_line_to_codes(Stream, Line),
    atom_codes(Atom, Line),
    atom_number(Atom, N),
    read_lines(Stream, N, Teams).

read_lines(Stream, N, Teams) :-
    ( N == 0 -> Teams = []
    ; N > 0  -> read_line(Stream, Team),
                Nm1 is N-1,
                read_lines(Stream, Nm1, RestTeams),
                Teams = [Team | RestTeams]).

read_line(Stream, team(Name, P, A, B)) :-
    read_line_to_codes(Stream, Line),
    atom_codes(Atom, Line),
    atomic_list_concat([Name | Atoms], ' ', Atom),
    maplist(atom_number, Atoms, [P, A, B]).

% [team(monaco, 2, 10, 2), team(andorra, 2, 11, 4), team(sanmarino, 1, 1, 4), team(liechtenstein, 1, 0, 7)]

beats(team(WName, WAgones, WEvale, WEfage), team(LName, LAgones, LEvale, LEfage)) :-
	WEvale > LEvale.

iswinner(team(Name, Agones, Evale, Efage)) :-
	Agones > 1.

isloser(team(Name, Agones, Evale, Efage)) :-
	Agones =:= 1.

split_teams(Teams, Winners, Losers) :-
	exclude(iswinner, Teams, Losers),
	exclude(isloser, Teams, Winners).

split_teams_finals(Teams, Winners, Losers) :-
	(
		Teams = [X, Y] -> Winners = [X], Losers = [Y];
		split_teams(Teams, Winners, Losers)
	).

valid([], []).
valid([W | Ws], [L | Ls]) :-
	beats(W, L), valid(Ws, Ls).

valid_round([], []).
valid_round(W, L, X) :-
	permutation(W, X),
	valid(X, L).

play_together([], [], [], []).
play_together(
	[team(WName, WAgones, WEvale, WEfage) | Ws],
 	[team(LName, LAgones, LEvale, LEfage) | Ls],
	[M | Matches], [NW | NewWinners]) :-
		NWAgones is WAgones - 1,
		NWEvale is WEvale - LEfage,
		NWEfage is WEfage - LEvale,
		M = match(WName, LName, LEfage, LEvale),
		NW = team(WName	, NWAgones, NWEvale, NWEfage),
		play_together(Ws, Ls, Matches, NewWinners).

% solve([team(WName, WAgones, WEvale, WEfage)], [team(LName, LAgones, LEvale, LEfage)], [Final]) :-
% 	WEvale >= 0, WEfage >= 0,
% 	LEvale >= 0, LEfage >= 0,
% 	WEvale =\= LEvale,
% 	Final = match(WName, LName, WEvale, LEvale).



solve(Winners, Losers, [M | Matches]) :-
  (
    Winners = [team(WName, WAgones, WEvale, WEfage)],
    Losers = [team(LName, LAgones, LEvale, LEfage)],
    WEvale >= 0, WEfage >= 0,
  	LEvale >= 0, LEfage >= 0,
  	WEvale =\= LEvale -> write('Le poule'), M = match(WName, LName, WEvale, LEvale), !;

    valid(Winners, Losers),
    play_together(Winners, Losers, M, TempWinners),
    write(M), nl,
    write('Temp Winners'),
    write(TempWinners), nl,
    once(split_teams(TempWinners, NewWinners, NewLosers)),
    write('New Winners'), nl,
    write('New Losers'), nl,
    write(NewWinners), nl,
    valid_round(NewWinners, NewLosers, CorrectWinners),
    solve(CorrectWinners, NewLosers, Matches)
  ).


mundial(File, Matches) :-
	read_input(File, N, Teams),
	split_teams(Teams, W, L),
	once(valid_round(W, L, InitWin)),
	write(InitWin), nl, write(L), nl,
	solve(InitWin, L, Matches).
